File: 69.sqrt-x.java
/*
 * @lc app=leetcode id=69 lang=java
 *
 * [69] Sqrt(x)
 */

// @lc code=start
class Solution {
    public int mySqrt(int x) {
        if (x < 2) { // Trả về x nếu x nhỏ hơn 2
            return x;
        }
        int left = 2, right = x / 2; // Thiết lập giới hạn tìm kiếm
        while (left <= right) { // Vòng lặp tìm kiếm nhị phân
            int mid = left + (right - left) / 2; // Tính giá trị giữa
            long square = (long) mid * mid; // Tính bình phương của mid
            if (square == x) { // Nếu bình phương bằng x, trả về mid
                return mid;
            } else if (square < x) { // Nếu bình phương nhỏ hơn x, cập nhật left
                left = mid + 1;
            } else { // Nếu bình phương lớn hơn x, cập nhật right
                right = mid - 1;
            }
        }
        return right; // Trả về giá trị gần nhất dưới căn bậc hai của x
    } 
}
// @lc code=end

File: 69.sqrt-x.py
#
# @lc app=leetcode id=69 lang=python3
#
# [69] Sqrt(x)
#

# @lc code=start
class Solution:
    def mySqrt(self, x: int) -> int:
        if x < 2:
            return x
        low, high = 2, x//2
        while low <= high:
            middle = low + (high - low) // 2
            square = middle * middle
            if square == x:
                return middle
            elif square < x:
                low = middle + 1
            else:
                high = middle - 1
        return high #asddS
# @lc code=end

